extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js

```diff
@@ -549,27 +541,69 @@ export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
   }
 
   /** @override */
-  getAdUrl() {
+  getAdUrl(opt_rtcResponsesPromise) {
     if (this.iframe && !this.isRefreshing) {
       dev().warn(TAG, `Frame already exists, sra: ${this.useSra}`);
       return '';
     }
+    opt_rtcResponsesPromise = opt_rtcResponsesPromise || Promise.resolve();
     // TODO(keithwrightbos): SRA blocks currently unnecessarily generate full
     // ad url.  This could be optimized however non-SRA ad url is required to
     // fallback to non-SRA if single block.
     this.populateAdUrlState();
     // TODO: Check for required and allowed parameters. Probably use
     // validateData, from 3p/3p/js, after noving it someplace common.
     const startTime = Date.now();
-    const rtcRequestPromise = isExperimentOn(this.win, 'disable-rtc') ?
-    Promise.resolve({}) : this.executeRtc_();
-    return rtcRequestPromise.then(rtcResult => {
+    return opt_rtcResponsesPromise.then(rtcResponseArray => {
+      const rtcParams = this.mergeRtcResponses_(rtcResponseArray);
       return googleAdUrl(
           this, DOUBLECLICK_BASE_URL, startTime, Object.assign(
-              this.getBlockParameters_(), rtcResult, PAGE_LEVEL_PARAMS_));
+              this.getBlockParameters_(), rtcParams, PAGE_LEVEL_PARAMS_));
+    });
+  }
+
+  /**
+   * Merges all of the rtcResponses into the JSON targeting and
+   * category exclusions.
+   * @param {?Array<rtcResponseDef>} rtcResponseArray
+   * @private
+   */
+  mergeRtcResponses_(rtcResponseArray) {
+    if (!rtcResponseArray) {
+      return null;
+    }
+    let rtcParams = {};
+    const artc = [];
+    const ati = [];
+    const ard = [];
+    rtcResponseArray.forEach(rtcResponse => {
+      // Only want to send errors for requests we actually sent.
+      if (rtcResponse.error &&
+          rtcResponse.error != RTC_ERROR_ENUM.MALFORMED_JSON_RESPONSE &&
```

Comment:
What about the timeout case?

Reply:
done, made it its own error case. It had previously been contained within network failure